equinix/resource_metal_device.go

```diff
@@ -726,33 +674,30 @@ func resourceMetalDeviceRead(ctx context.Context, d *schema.ResourceData, meta i
 }
 
 func resourceMetalDeviceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
-	meta.(*config.Config).AddModuleToMetalUserAgent(d)
-	client := meta.(*config.Config).Metal
+	meta.(*config.Config).AddModuleToMetalGoUserAgent(d)
+	client := meta.(*config.Config).Metalgo
+
+	ur := metalv1.DeviceUpdateInput{}
 
 	if d.HasChange("locked") {
```

Comment:
So...after digging into what it means for a device to be locked and confirming that we don't have to use a separate update request to change the lock status, it turns out that `locked` is a computed attribute for a `metal_device` resource, so it's not actually possible to hit this code path anyway.

So:  do we want to stick with the existing behavior that `locked` can't be changed or do we want to support locking/unlocking devices in terraform?

For reference, this is the information we got about what locking does:
> It seems that locking a device prevents you from:
> - Delete/deprovision the device
> - Reinstall the device
> - Prevents an instance with a termination time set to be reclaimed, even if the termination time was reached (for spot instances, we will reclaim it anyway)
> - Perform a Firmware update on the device (this feature is not GA yet)
Perform a Firmware update on the device (this feature is not GA yet)

Reply:
I feel like it's a core state for our devices to support and so terraform should be able to do it.

Reply:
The list of lock affecting features is helpful context. Without that we may expect behaviors that are not present today: https://feedback.equinixdigital.com/instances/p/expand-lock-feature-for-metal-servers.

It's worth considering the ways Terraform and a Portal user would interact through [EM API device locking](https://metal.equinix.com/developers/api/#tag/Devices/operation/updateDevice) (not to be confused with [Terraform state locking](https://developer.hashicorp.com/terraform/language/state/locking)).

If Terraform takes an "optional" (default=false) behavior, rather than computed, then plan/apply from Terraform would inform users that an externally applied `lock` is going to be reset upon change.

When Terraform config requests `lock`, then any API changes that are incompatible with `lock` would fail until the Terraform config is modified to `lock=false` (or the `lock` line is removed).

I'm not seeing any problems in taking this approach.

Reply:
OK, sounds like we're all agreed; I'm going to update the resource so that `locked` can be set and update the docs accordingly.

Reply:
I enabled locking/unlocking a device resource; this PR has been updated but here's a direct link to the commit that has those changes: https://github.com/equinix/terraform-provider-equinix/pull/501/commits/836f1f045e284accc17f47a4801acc4b45f9726f

I opted to make `locked` optional and computed, rather than optional with a default.  IMO changing it from computed to default has a bigger impact on existing users, because if we switch to a default of false then any terraform user who currently has a locked device is at risk for accidentally unlocking the device as a result of a non-major-version provider upgrade.